# R-Studio Time Series

## 3. Univariate Time Series: ARIMA, ARCH, dan GARCH

In [11]:
View(hargaberas)

## 3.1. Box-Jenkins

Prosedur Box-Jenkins dalam ARIMA
1. Identifikasi
2. Estimasi Model
3. Evaluasi Model
4. Peramalan

## 3.1. Identifikasi

### 3.1.2. Inisialisasi Data

In [9]:
library(readxl)
hargaberas <- read_excel('C:\\Users\\hisbi\\iCloudDrive\\repository\\R-tutorial\\data-i\\Bab 3\\ARIMA.xlsx')

### 3.1.2. Inisialisasi Data Time Series

In [12]:
hargaberas = hargaberas[,c(-1)]
hargaberas = ts(hargaberas, start=c(2012,1), frequency=12)
hargaberas

       Jan   Feb   Mar   Apr   May   Jun   Jul   Aug   Sep   Oct   Nov   Dec
2012  8726  8778  8687  8583  8537  8554  8606  8635  8624  8624  8655  8702
2013  8835  8843  8783  8711  8681  8784  9018  9057  9058  9108  9152  9262
2014  9433  9531  9596  9425  9414  9462  9525  9525  9694  9781  9924 10344
2015 10612 10766 10987 10648 10569 10679 10732 10935 11055 11169 11365 11465
2016 11614 11729 11678 11449 11417 11469 11498 11475 11448 11433 11450 11476
2017 11579 11571 11494 11449 11465 11465 11448 11411 11482 11552 11665 11838
2018 12276 12414 12299 12035 11943 11907 11936 11899 11900 11926 12013 12106


[,c(-1)] = Menghapus kolom ke-1 pada objek harga beras.

ts = Membuat periode waktu data harga beras (format sebelumnya belum dapat dibaca oleh software RStudio)
data = 
start = observasi pertama pada data, dapat berupa nomor tunggal atau vektor
frequency = jumlah observasi per unit waktu (kuartal = 4, tahun = 1), tidak cocok digunakan untuk data harian karena biasanya terdapat perbedaan unit waktu dalam data harian.

### 2.1.3. Visualisasi Harga Beras

In [13]:
par(mfrow=c(1,1))
plot(hargaberas, main="Harga Beras di Perdagangan Besar")

### 2.1.4. Dekomposisi

Fungsi dekomposisi digunakan untuk melakukan dekomposisi data hargaberas menjadi beberapa komponen time series

1. random 
2. seasonal
3. trend
4. observed


Hal ini berguna untuk mengetahui keberadaan komponen time series sehingga menjadi pertimbangan dalam membuat model

In [19]:
dekomposisi = decompose(hargaberas)
plot(dekomposisi)

### 3.1.5. Melihat ACF dan PACF

In [35]:
par (mfrow=c(2, 1)) # Jendela plot 2 baris 1 kolom
acf(hargaberas, lag=48)
pacf(hargaberas, lag=48)

: Error in `plot.new()`:
! figure margins too large

### 3.1.6. Uji Stationeritas

In [38]:
library("aTSA")
adf.test(hargaberas)


Attaching package: ‘aTSA’

The following object is masked from ‘package:graphics’:

    identify



Attaching package: ‘aTSA’

The following object is masked from ‘package:graphics’:

    identify



Augmented Dickey-Fuller Test 
alternative: stationary 
 
Type 1: no drift no trend 
     lag  ADF p.value
[1,]   0 2.96   0.990
[2,]   1 1.65   0.975
[3,]   2 2.04   0.990
[4,]   3 2.18   0.990
Type 2: with drift no trend 
     lag    ADF p.value
[1,]   0 -0.537   0.858
[2,]   1 -0.717   0.795
[3,]   2 -0.831   0.755
[4,]   3 -0.917   0.724
Type 3: with drift and trend 
     lag   ADF p.value
[1,]   0 -1.42   0.814
[2,]   1 -2.56   0.340
[3,]   2 -2.00   0.570
[4,]   3 -1.78   0.663
---- 
Note: in fact, p.value = 0.01 means p.value <= 0.01 
Augmented Dickey-Fuller Test 
alternative: stationary 
 
Type 1: no drift no trend 
     lag  ADF p.value
[1,]   0 2.96   0.990
[2,]   1 1.65   0.975
[3,]   2 2.04   0.990
[4,]   3 2.18   0.990
Type 2: with drift no trend 
     lag    ADF p.value
[1,]   0 -0.537   0.858
[2,]   1 -0.717   0.795
[3,]   2 -0.831   0.755
[4,]   3 -0.917   0.724
Type 3: with drift and trend 
     lag   ADF p.value
[1,]   0 -1.42   0.814
[2,]   1 -2.56   0.340
[3,]   2 -2

In [39]:
adf.test(diff(hargaberas))

Augmented Dickey-Fuller Test 
alternative: stationary 
 
Type 1: no drift no trend 
     lag   ADF p.value
[1,]   0 -5.11    0.01
[2,]   1 -5.02    0.01
[3,]   2 -4.42    0.01
[4,]   3 -4.36    0.01
Type 2: with drift no trend 
     lag   ADF p.value
[1,]   0 -5.47    0.01
[2,]   1 -5.58    0.01
[3,]   2 -5.10    0.01
[4,]   3 -5.27    0.01
Type 3: with drift and trend 
     lag   ADF p.value
[1,]   0 -5.43    0.01
[2,]   1 -5.55    0.01
[3,]   2 -5.10    0.01
[4,]   3 -5.28    0.01
---- 
Note: in fact, p.value = 0.01 means p.value <= 0.01 
Augmented Dickey-Fuller Test 
alternative: stationary 
 
Type 1: no drift no trend 
     lag   ADF p.value
[1,]   0 -5.11    0.01
[2,]   1 -5.02    0.01
[3,]   2 -4.42    0.01
[4,]   3 -4.36    0.01
Type 2: with drift no trend 
     lag   ADF p.value
[1,]   0 -5.47    0.01
[2,]   1 -5.58    0.01
[3,]   2 -5.10    0.01
[4,]   3 -5.27    0.01
Type 3: with drift and trend 
     lag   ADF p.value
[1,]   0 -5.43    0.01
[2,]   1 -5.55    0.01
[3,]   2 -5

## 3.2. Estimasi Model

In [40]:
library(forecast)
auto.arima(hargaberas, trace=TRUE)

Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo 
Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo 



Attaching package: ‘forecast’

The following object is masked from ‘package:aTSA’:

    forecast



Attaching package: ‘forecast’

The following object is masked from ‘package:aTSA’:

    forecast



ARIMA(2,1,2)(1,1,1)[12]                    : Inf
 ARIMA(0,1,0)(0,1,0)[12]                    : 884.0934
 ARIMA(1,1,0)(1,1,0)[12]                    : 872.3993
 ARIMA(0,1,1)(0,1,1)[12]                    : Inf
 ARIMA(1,1,0)(0,1,0)[12]                    : 876.5573ARIMA(2,1,2)(1,1,1)[12]                    : Inf
 ARIMA(0,1,0)(0,1,0)[12]                    : 884.0934
 ARIMA(1,1,0)(1,1,0)[12]                    : 872.3993
 ARIMA(0,1,1)(0,1,1)[12]                    : Inf
 ARIMA(1,1,0)(0,1,0)[12]                    : 876.5573


 ARIMA(1,1,0)(2,1,0)[12]                    : 863.6643

 ARIMA(1,1,0)(2,1,0)[12]                    : 863.6643


 ARIMA(1,1,0)(2,1,1)[12]                    : Inf
 ARIMA(1,1,0)(1,1,1)[12]                    : Inf
 ARIMA(0,1,0)(2,1,0)[12]                    : 867.3178

 ARIMA(1,1,0)(2,1,1)[12]                    : Inf
 ARIMA(1,1,0)(1,1,1)[12]                    : Inf
 ARIMA(0,1,0)(2,1,0)[12]                    : 867.3178


 ARIMA(2,1,0)(2,1,0)[12]                    : 865.871

 ARIMA(2,1,0)(2,1,0)[12]                    : 865.871


 ARIMA(1,1,1)(2,1,0)[12]                    : 865.1084
 ARIMA(0,1,1)(2,1,0)[12]                    : 862.9856


 ARIMA(1,1,1)(2,1,0)[12]                    : 865.1084
 ARIMA(0,1,1)(2,1,0)[12]                    : 862.9856


 ARIMA(0,1,1)(1,1,0)[12]                    : 871.9363 ARIMA(0,1,1)(1,1,0)[12]                    : 871.9363


 ARIMA(0,1,1)(2,1,1)[12]                    : Inf
 

 ARIMA(0,1,1)(2,1,1)[12]                    : Inf
 

ARIMA(0,1,1)(1,1,1)[12]                    : InfARIMA(0,1,1)(1,1,1)[12]                    : Inf


 ARIMA(0,1,2)(2,1,0)[12]                    : 861.8543
 ARIMA(0,1,2)(1,1,0)[12]                    : 872.2305

 ARIMA(0,1,2)(2,1,0)[12]                    : 861.8543
 ARIMA(0,1,2)(1,1,0)[12]                    : 872.2305


 ARIMA(0,1,2)(2,1,1)[12]                    : Inf


 ARIMA(0,1,2)(2,1,1)[12]                    : Inf


 ARIMA(0,1,2)(1,1,1)[12]                    : Inf ARIMA(0,1,2)(1,1,1)[12]                    : Inf


 ARIMA(1,1,2)(2,1,0)[12]                    : 867.0498

 ARIMA(1,1,2)(2,1,0)[12]                    : 867.0498


 ARIMA(0,1,3)(2,1,0)[12]                    : 854.8026
 

 ARIMA(0,1,3)(2,1,0)[12]                    : 854.8026
 

ARIMA(0,1,3)(1,1,0)[12]                    : 865.2557ARIMA(0,1,3)(1,1,0)[12]                    : 865.2557


 ARIMA(0,1,3)(2,1,1)[12]                    : Inf

 ARIMA(0,1,3)(2,1,1)[12]                    : Inf


 ARIMA(0,1,3)(1,1,1)[12]                    : Inf

 ARIMA(0,1,3)(1,1,1)[12]                    : Inf


 ARIMA(1,1,3)(2,1,0)[12]                    : Inf

 ARIMA(1,1,3)(2,1,0)[12]                    : Inf


 ARIMA(0,1,4)(2,1,0)[12]                    : 857.2177

 ARIMA(0,1,4)(2,1,0)[12]                    : 857.2177


 ARIMA(1,1,4)(2,1,0)[12]                    : 859.7882

 Best model: ARIMA(0,1,3)(2,1,0)[12]                    



 ARIMA(1,1,4)(2,1,0)[12]                    : 859.7882

 Best model: ARIMA(0,1,3)(2,1,0)[12]                    



Series: hargaberas 
ARIMA(0,1,3)(2,1,0)[12] 

Coefficients:
         ma1     ma2     ma3     sar1     sar2
      0.3775  0.0028  0.4180  -0.4831  -0.4956
s.e.  0.1166  0.1301  0.1403   0.1231   0.1239

sigma^2 = 7757:  log likelihood = -420.75
AIC=853.49   AICc=854.8   BIC=867.07

Berdasarkan output tersebut, model terbaik yang menjadi rekomendasi adalah ARIMA (0,1,3)(2,1,0)[12]

Hal tersebut karena memiliki AIC terkecil

Model tersebut juga
1. Memiliki komponen nonseasonal
2. Seasonal

In [46]:
# Mengetahui nilai koefisien, standard error, dan p-value
library(lmtest)
# ARIMA(0,1,3)(2,1,0)[12]
model1 = arima(hargaberas, order=c(0,1,3), seasonal = list(order=c(2,1,0), period=12)) # Model ARIMA manual
coeftest(model1)

Loading required package: zoo

Attaching package: ‘zoo’

The following objects are masked from ‘package:base’:

    as.Date, as.Date.numeric

Loading required package: zoo

Attaching package: ‘zoo’

The following objects are masked from ‘package:base’:

    as.Date, as.Date.numeric




z test of coefficients:

      Estimate Std. Error z value  Pr(>|z|)    
ma1   0.377526   0.116627  3.2370  0.001208 ** 
ma2   0.002799   0.130115  0.0215  0.982837    
ma3   0.417985   0.140264  2.9800  0.002883 ** 
sar1 -0.483126   0.123055 -3.9261 8.634e-05 ***
sar2 -0.495630   0.123883 -4.0008 6.313e-05 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


Model1 merupakan model dengan AIC terkecil dan juga parsimoniuos. Koefisien dugaan nyata atay signidikan dan tellah memenuhi kondisi invertibilitas

Invertibilitas = Nilai koefisien kurang dari satu

## 3.3. Evaluasi Model

Uji Heteroskedastisitas dan keacakan residual

### 3.3.1. Uji Heteroskedastisitas

In [47]:
# Uji heteroskedastisitas --> Cukup arima atau perlu dilanjutkan dengan ARCH/GARCH
arch.test(model1)

ARCH heteroscedasticity test for residuals 
alternative: heteroscedastic 

Portmanteau-Q test: 
     order    PQ p.value
[1,]     4  5.41   0.247
[2,]     8  8.53   0.384
[3,]    12 11.36   0.498
[4,]    16 13.03   0.670
[5,]    20 14.44   0.808
[6,]    24 16.04   0.887
Lagrange-Multiplier test: 
     order    LM  p.value
[1,]     4 41.16 6.04e-09
[2,]     8 13.84 5.42e-02
[3,]    12  6.06 8.69e-01
[4,]    16  3.39 9.99e-01
[5,]    20  2.16 1.00e+00
[6,]    24  1.16 1.00e+00
ARCH heteroscedasticity test for residuals 
alternative: heteroscedastic 

Portmanteau-Q test: 
     order    PQ p.value
[1,]     4  5.41   0.247
[2,]     8  8.53   0.384
[3,]    12 11.36   0.498
[4,]    16 13.03   0.670
[5,]    20 14.44   0.808
[6,]    24 16.04   0.887
Lagrange-Multiplier test: 
     order    LM  p.value
[1,]     4 41.16 6.04e-09
[2,]     8 13.84 5.42e-02
[3,]    12  6.06 8.69e-01
[4,]    16  3.39 9.99e-01
[5,]    20  2.16 1.00e+00
[6,]    24  1.16 1.00e+00


Berdasarkan output, tidak terdapat heteroskedastisitas pada model.

### 3.3.2. Uji Keacakan Reisudal

In [48]:
# diagnostik mmodel (autokorelasi)
Box.test(model1$residuals, lag = 1, type = c("Ljung-Box"), fitdf = 0)


	Box-Ljung test

data:  model1$residuals
X-squared = 0.0041444, df = 1, p-value = 0.9487


Berdasarkan output, p-value 0.9487 > taraf signifikansi (0.05) sehingga dapat disimpulkan gagal tolak H0, sehingga tidak cukup bukti untuk menyatakan adanya autokorelasi.

Seluruh asumsi pada evaluasi model terpenuhi sehingga dapat melakukan peramalan data harga beras dengan model 1

## 3.4. Forecast

In [49]:
forecast(model1, h=12)

         Point Forecast    Lo 80    Hi 80    Lo 95    Hi 95
Jan 2019       12355.67 12246.95 12464.39 12189.39 12521.95
Feb 2019       12493.09 12308.02 12678.16 12210.05 12776.13
Mar 2019       12392.80 12154.53 12631.07 12028.40 12757.21
Apr 2019       12143.41 11835.19 12451.63 11672.03 12614.80
May 2019       12079.80 11714.79 12444.80 11521.57 12638.02
Jun 2019       12086.96 11672.89 12501.04 11453.70 12720.23
Jul 2019       12116.54 11658.63 12574.45 11416.22 12816.85
Aug 2019       12086.48 11588.57 12584.38 11325.00 12847.96
Sep 2019       12072.73 11537.81 12607.64 11254.64 12890.81
Oct 2019       12077.85 11508.32 12647.38 11206.83 12948.88
Nov 2019       12129.83 11527.68 12731.99 11208.92 13050.75
Dec 2019       12188.63 11555.52 12821.73 11220.38 13156.88

In [55]:
plot(forecast(model1, h=12))

: Error in `plot.new()`:
! figure margins too large